src/cursor/cur_std.c

```diff
@@ -622,6 +634,61 @@ __wt_cursor_set_value(WT_CURSOR *cursor, ...)
     va_end(ap);
 }
 
+/*
+ * __wt_cursor_set_raw_key_value --
+ *     WT_CURSOR->set_raw_key_value default implementation
+ */
+void
+__wt_cursor_set_raw_key_value(WT_CURSOR *cursor, WT_ITEM *key, WT_ITEM *value)
+{
+    WT_DECL_RET;
+    WT_ITEM *buf;
+    WT_SESSION_IMPL *session;
+
+    CURSOR_API_CALL(cursor, session, ret, set_raw_key_value, NULL);
+    WT_ERR(__cursor_copy_release(cursor));
```

Comment:
Did you consider refactoring and using `set_raw_key` and `set_raw_value` internally? I know we are chasing performance here, but the duplication of complex and subtle code worries me.

There also doesn't seem to be enough checking to make sure the API is used as expected. I'd expect that to be present in at least diagnostic builds, and preferably everywhere: This is a primary mechanism for applications to give context to WiredTiger, which it will operate on. Ensuring the data is in an expected state/format on entry is necessary to limit how much further checking/safety we need inside the library.

Reply:
Hi @agorrod , as I understand it, part of the challenge here is that `__wt_cursor_set_raw_key` and `__wt_cursor_set_raw_value` call `set_key` and `set_value` respectively, and those functions take va args parameters. Removing the va args usage (and perf cost) is a main benefit from this proposed change. 
Obviously, if there are opportunities to refactor the code and/or take a simpler approach then we should do so.